Rewind IO stream before computing its checksum

matthieuprat opened the pull request:
### Background

When Active Storage uploads a stream, it does the following:

  1. Compute a checksum of the stream[^1]
  2. Call the storage service to perform the actual upload
  3. Compare the checksum computed in step 1 with the checksum of the uploaded file and raise an `IntegrityError` if they don't match[^2]

The checksum in step 1 is computed from the current offset of the stream but the stream is rewound after. Thus, the whole stream is uploaded in step 2, regardless of the stream's initial offset. As a result, if the stream has an initial offset, the integrity check in step 3 fails.

A few people have run into this issue[^3] and the root cause is not immediately obvious from the error raised by Active Storage.

### Solution

If the stream is at end-of-stream (i.e. if `io.eof?` is true), then the user most likely expects the whole stream to be uploaded. So let's rewind the stream in that case.

If the stream has an offset but is not at end-of-stream, then the user could arguably expect the stream to be uploaded from the offset. In that case, rewinding the stream would be surprising behaviour. Therefore, let's preserve the existing behaviour in this scenario, i.e. raise an `IntegrityError`. Ideally, we would bail out earlier with a more helpful error but I wasn't sure where to raise so I'm leaving this as a follow-up.

### Checklist

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message.
* [x] Tests are added or updated if you fix a bug or add a feature.



[^1]: https://github.com/rails/rails/blob/v8.0.0/activestorage/app/models/active_storage/blob.rb#L348-L359
[^2]: https://github.com/rails/rails/blob/v8.0.0/activestorage/lib/active_storage/service/disk_service.rb#L163-L168
[^3]: https://stackoverflow.com/questions/52335835/when-is-activestorageintegrityerror-raised

byroot commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -349,6 +349,7 @@ def compute_checksum_in_chunks(io)
       raise ArgumentError, "io must be rewindable" unless io.respond_to?(:rewind)
 
       OpenSSL::Digest::MD5.new.tap do |checksum|
+        io.rewind
```
I think we should save `IO#pos`, and reset to that offset. Otherwise it will also fail is you pass an IO with a non-zero offset.

matthieuprat replied:
Thank you for taking a look!

I think it works for partially-consumed streams but the entire stream is uploaded, which is maybe not what one would expect. I added a test to demonstrate this (5b3f41a).

I also tried your suggestion of preserving the offset (e76bc70). It seems to work with the disk service but I'm not sure it will for others. For instance, it seems the Ruby client library for Google Cloud Storage rewinds the stream before the upload, so we'd end up getting an integrity error.

Thus, I'm wondering if we wouldn't be better off either always rewinding or bailing out early with an helpful error for partially-consumed streams. Do you have a steer?

byroot replied:
>  For instance, it seems the Ruby client library for Google Cloud Storage rewinds the stream before the upload, so we'd end up getting an integrity error.

Argh, that's really unfortunate. I went to check the code, and indeed, it doesn't care about offsets.

> Do you have a steer?

I'm not the most familiar with Active Storage, but we probably should try to be consistent. So you initial rewinf is probably the way to go, but my worry is some people might have been relying on that behavior on some backends.

So we should add a CHANGELOG etc.

matthieuprat replied:
To clarify, I think the behavior is currently consistent across all backends: when given a stream with an offset, Active Storage raises an `IntegrityError` (adb6f69).

That's because Active Storage rewinds the stream before handing it over to the backend, namely right after the checksum is calculated: https://github.com/rails/rails/blob/v8.0.0/activestorage/app/models/active_storage/blob.rb#L357

So rewinding the stream shouldn't be breaking, I think? Although I appreciate that the resulting behavior (i.e. uploading the whole stream) might be surprising if the stream has an offset and is not at EOF. In that case, maybe we're better off raising an error. Do you have any thoughts on  d25676c?

byroot replied:
> Do you have any thoughts on [d25676c](https://github.com/rails/rails/commit/d25676c535c6fb11c2ec61e9f42fe034c1f501dc)?

I think it's a good idea. But I'm not a big user of Active Storage so no strong opinion.

matthieuprat replied:
OK, thank you again for your time and feedback!

matthieuprat replied:
@jonathanhefner, I hope you don’t mind the mention. I noticed you contributed to Active Storage recently-ish and thought you might be well-positioned to have an opinion on this pull request. I’d appreciate your thoughts if you have the time. Thank you!

matthieuprat committed:
Add tests for uploading streams with an offset

The `ActiveStorage::Blob.create_and_upload!` method raises an
`IntegrityError` when given an IO stream that has an offset.

This commit adds tests to show this unexpected behavior.

matthieuprat committed:
Support uploading a stream which is at end-of-stream

When Active Storage uploads a stream, it does the following:

  1. Compute a checksum of the stream[^1]
  2. Call the storage service to perform the actual upload
  3. Compare the checksum computed in step 1 with the checksum of
     the uploaded file and raise an `IntegrityError` if they don't
     match[^2]

The checksum in step 1 is computed from the current offset of the stream
but the stream is rewound after. Thus, the whole stream is uploaded in
step 2, regardless of the stream's initial offset. As a result, if the
stream has an initial offset, the integrity check in step 3 fails.

A few people have run into this issue[^3] and the root cause is not
immediately obvious from the error raised by Active Storage.

If the stream is at end-of-stream (i.e. if `io.eof?` is true), then the
user most likely expects the whole stream to be uploaded. So let's
rewind the stream in that case.

If the stream has an offset but is not at end-of-stream, then the user
could arguably expect the stream to be uploaded from the offset. In that
case, rewinding the stream would be surprising behaviour. Therefore,
let's preserve the existing behaviour in this scenario, i.e. raise an
`IntegrityError`. Ideally, we would bail out earlier with a more helpful
error but I wasn't sure where to raise so I'm leaving this as a
follow-up.

[^1]: https://github.com/rails/rails/blob/v8.0.0/activestorage/app/models/active_storage/blob.rb#L348-L359
[^2]: https://github.com/rails/rails/blob/v8.0.0/activestorage/lib/active_storage/service/disk_service.rb#L163-L168
[^3]: https://stackoverflow.com/questions/52335835/when-is-activestorageintegrityerror-raised
